Add ``.address`` and ``.selector`` in inside assembly for external function pointers (#12016)
ethereum/solidity: docs/assembly.rst

The variable can also be assigned a new offset, but note that no validation to ensure that
the variable will not point beyond ``calldatasize()`` is performed.

For external function pointers the address and the function selector can be
accessed using ``x.address`` and ``x.selector``.

> The selector will be left aligned and of type ``uint32``.

Reviewer: At first I read that as if the selector was being left somewhere :)
```suggestion
The selector will be left-aligned and of type ``uint32``.
```

> The selector will be left aligned and of type ``uint32``.

Reviewer: If there's no unused space then isn't it technically both left- and right-aligned? Or not aligned at all? I'm not sure if it makes sense to speak of alignment of a 4-byte value in a 4-byte type.

Reviewer: Ah, it probably means alignment in the 32-byte slot. I think it should be clarified here. But then I do not see in what sense it's a `uint32` value - the type depends on the type of an underlying variable.

Reviewer 2: There are no types in assembly, so I would propose to talk about four bytes that are left-aligned or something like that.

> Assigning is not possible.

Reviewer: How about this?
```suggestion
Both values are read-only and cannot be assigned.
```

The variable can also be assigned a new offset, but note that no validation to ensure that
the variable will not point beyond ``calldatasize()`` is performed.

For external function pointers the address and the function selector can be
accessed using ``x.address`` and ``x.selector``.
The selector consists of four right-aligned bytes.
Both values are can be assigned to. For example:

.. code-block:: solidity
    :force:

    // SPDX-License-Identifier: GPL-3.0
    pragma solidity >=0.8.10 <0.9.0;

>     contract C {
>         function externalFunction() external {}
> 
>         // Assigns a new selector and address to the local variable fp
>         function assignSelector(address newAddress, uint newSelector) view public {
>             function() external fp = this.externalFunction;
> 
>             assembly {
>                 fp.selector := newSelector
>                 fp.address  := newAddress
>             }
> 
>             assert(fp.address == newAddress);
>             assert(fp.selector == bytes4(bytes32(newSelector)));
>         }
>     }

Reviewer 2: What about (please check that it compiles)
```suggestion
    // Assigns a new selector and address to the local variable fp
    function combineToFunctionPointer(address newAddress, uint newSelector) pure returns (function() external fun)
            assembly {
                fun.selector := newSelector
                fun.address  := newAddress
            }
        }
    }
```
